### PR TITLE
python: add FutureExt class for extensible futures from python

### PR DESCRIPTION
**Pull request by grondo**
Problem: There is no current support for easily building extensible Future objects in the Python API, e.g. to return a Future that may be composed of a series of multiple RPCs or other actions with a single result.

Add a `FutureExt` class, which is derived from `Future`, but takes an initialization callback like `flux_future_t` which can be used to initiate the composite series of actions that should result in fulfillment of the returned `Future`. The `FutureExt` class also support fulfilling a Future with any Python object that is JSON serializable, and the `FutureExt.get()` method is overloaded to return the fulfilled value. This makes it simple to return a result from Python to Python using `FutureExt`.

A simple example is used in testing and copied here for demonstration purposes: a `Ping5` future which is fulfilled after 5 successful pings, used in both synchronous and async contexts:

```python
import flux
from flux.future import FutureExt


class Ping5(FutureExt):
    def __init__(self, flux_handle):
        super().__init__(init_cb=self.init_cb, flux_handle=flux_handle)

    def ping_cb(self, future):
        seq = future.get()["seq"]
        if seq == 5:
            try:
                self.fulfill(future.get())
            except Exception as exc:
                print(exc)
            return
        h = future.get_flux()
        h.rpc("broker.ping", {"seq": seq + 1}).then(self.ping_cb)

    def init_cb(self, future):
        h = future.get_flux()
        h.rpc("broker.ping", {"seq": 1}).then(self.ping_cb)


print(Ping5(flux.Flux()).get())

h = flux.Flux()
Ping5(h).then(lambda future: print(future.get()))
h.reactor_run()
```

```console
$ python3 ping5.py
{'seq': 5, 'route': '7240baa1!c201314c!a2da1fa8', 'userid': 1000, 'rolemask': 5, 'rank': 0}
{'seq': 5, 'route': '12776d3e!c201314c!a2da1fa8', 'userid': 1000, 'rolemask': 5, 'rank': 0}
```

The main use of this class for now though will be to handle an API for job output, which will be forthcoming in a future PR.

### CODE REVIEW

**Comment by grondo** [comment]
Should have mentioned there's also some fixes and improvements for the existing futures stuff that popped up when testing the `FutureExt` interface.

One bit issue that future developers will hit is that when an exception is raised in a Python callback, and that causes the callback to return without fulfilling the future on which a `get()` or `wait_for()` was called, libflux-core considers this an error and returns `EDEADLOCK`. Commit [42d9974](https://github.com/flux-framework/flux-core/pull/5330/commits/42d997438a30999e9d5b535c81aaf362a09a125e) handles this by checking for any pending exception attribute set in the `Flux` class for the current thread, and raises this exception instead of `OSError` with `EDEADLOCK` (which is unhelpful). This only became a problem with the `FutureExt` class because this error only occurs if the Python callback is responsible for fulfilling a future, which wasn't possible before. I just thought this could use some extra explanation since it took me a fair bit of time to run down what was happening here.

**Review by garlick — APPROVED** [review]
Hey that turned out pretty neat!

I didn't spot anything wrong here and I know you've already built things on top of this and worked through some issues.  Might be good to get another pair of eyes on the python part, but FWIW LGTM!

File: src/bindings/python/flux/future.py [hunk 1]
```diff
@@ -243,3 +268,113 @@ def push(self, child, name=None):
...
+        The initialization callback should take care to use
+        ``future.get_flux()`` to get the correct Flux handle for the
+        context in which the initialization is running. This is because
+        the underlying reactor will be different depending on whether
+        the Future is being initialized in a blocking ``future.get()``
+        or by a call to ``future.then()``. See the documentation for
+        ``flux_future_create(3)`` for more information.
+
+        If ``flux_handle`` is not None, then the give Flux handle will
+        be associated with the Future during initialization. Otherwise,
```
Comment (chu11) [inline src/bindings/python/flux/future.py:303]: give -> given?

File: t/python/t0012-futures.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 import flux
 import flux.constants
 from flux.core.inner import ffi
-from flux.future import Future
+from flux.future import Future, FutureExt
```
Comment (chu11) [inline t/python/t0012-futures.py:20]: commit message typo "PYthon" :-)

**Comment by grondo** [comment]
Thanks! I addressed those two typos and forced a push.

File: t/python/t0012-futures.py [hunk 3]
```diff
@@ -240,6 +240,79 @@ def test_08_future_from_future(self):
...
+            self.assertEqual(a, "foo")
+            self.assertEqual(b, "bar")
+            cb_ran[0] = True
+            # XXX: reactor doesn't exit without this for unknown reason
+            # works standalone though...
+            future.get_flux().reactor_stop()
+
+        Ping5(self.f).then(ping5_cb, "foo", b="bar")
+        self.f.reactor_run()
+
```
Comment (chu11) [inline t/python/t0012-futures.py:315]: don't check cb_ran[0] == true?

**Review by chu11 — APPROVED** [review]
LGTM, just noticed a few small things

File: t/python/t0012-futures.py [hunk 4]
```diff
@@ -240,6 +240,79 @@ def test_08_future_from_future(self):
...
+        def init_cb(future):
+            cb_ran[0] = True
+            self.assertIsInstance(future, Future)
+            future.fulfill("foo")
+
+        future = FutureExt(init_cb)
+        self.assertIsInstance(future, FutureExt)
+        self.assertEqual(future.get(), "foo")
+        self.assertTrue(cb_ran[0])
```
[the remark this answers is not in the document]
Comment (grondo) [reply t/python/t0012-futures.py:254]: Hm, just curious why? It is initialized to `False` and this check isn't done in any other similar tests in the testsuite.

[the remark this answers is not in the document]
Comment (chu11) [reply t/python/t0012-futures.py:254]: I figured just to make sure `init_cb` isn't called before the call to get(), which I don't think it should be, right?

[the remark this answers is not in the document]
Comment (grondo) [reply t/python/t0012-futures.py:254]: Ah, ok. You want to ensure that the intitialization callback isn't called just be instantiating the `FutureExt` object. I guess that is fair and certainly doesn't hurt.

[the remark this answers is not in the document]
Comment (grondo) [reply t/python/t0012-futures.py:254]: Note: a similar argument could be made that the tests should verify that `cb_ran == False` after calling `.then()` on a Future and before entering the reactor, but that seems even more excessive than the check here.

**Comment by grondo** [comment]
Ok force-pushed suggested changes.

File: t/python/t0012-futures.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chu11) [reply t/python/t0012-futures.py]: Yeah, that seems excessive.  Which is why I didn't mention it below, figured this was enough :-)

[the remark this answers is not in the document]
Comment (grondo) [reply t/python/t0012-futures.py]: Heh. I guess I was trying to say I saw it as the same level of excessive 😜

**Comment by grondo** [comment]
restarting fedora build which failed with known issue
```
Error: ERROR: test_remote.t - missing test plan
Error: ERROR: test_remote.t - exited with status 137 (terminated by signal 9?)
```

**Comment by grondo** [comment]
With help from @trws I found an issue with the python `Future` ffi handle that was causing much grief in work using the `FutureExt` class. I've dropped 2 extra commits here with an explanation. I wasn't able to come up with reproducer for the `python/t0012-futures.py` test, but these changes to handling of the `Future` ffi handles seems to be necessary for some of the job output watching classes. I put the commits here though since it more logically seemed to go with this PR. 

I could also add them in a follow on PR if that's preferable.

**Comment by grondo** [comment]
Hm, since this PR was already approved without the last two commits. Maybe I'll remove those, merge this, and then propose the workaround for the series of unfortunately garbage collected Future ffi handles as a separate PR.

**Comment by grondo** [comment]
Setting MWP.

**Pull request merged**